### PR TITLE
feat: add MemorystoreDocumentSaver class

### LINKED ISSUE #14: Implement document saver

### PR DESCRIPTION
**Pull request by craig08**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/langchain-google-memorystore-redis-python/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #14 🦕

### CODE REVIEW

File: src/langchain_google_memorystore_redis/__init__.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 from langchain_google_memorystore_redis.chat_message_history import MemorystoreChatMessageHistory
+from langchain_google_memorystore_redis.doc_saver import MemorystoreDocumentSaver
```
Comment (PingXie) [inline src/langchain_google_memorystore_redis/__init__.py:15]: Can we use the relative path instead? it is more resilient to potential name changes in the future

File: src/langchain_google_memorystore_redis/doc_saver.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+class MemorystoreDocumentSaver:
+    """Document Saver for Cloud Memorystore for Redis database."""
+
+    def __init__(
+        self,
+        client: redis.Redis,
+        content_field: str,
+        metadata_fields: Optional[Set[str]] = None,
+        key_prefix: Optional[str] = None,
```
Comment (PingXie) [inline src/langchain_google_memorystore_redis/doc_saver.py:31]: should we make this mandatory - otherwise, with the current vector store implementation, I don't think we could create a vector store against documents saved this way.

File: src/langchain_google_memorystore_redis/doc_saver.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        """
+        doc_ids = ids
+        if not doc_ids:
+            doc_ids = [self._key_prefix + str(uuid.uuid4()) for _ in documents]
+        if len(documents) != len(doc_ids):
+            raise ValueError("The length of documents must match the length of the IDs")
+
+        for i, doc in enumerate(documents):
+            mapping = self._filter_metadata_by_fields(doc.metadata)
+            mapping.update({self._content_field: json.dumps(doc.page_content)})
```
Comment (PingXie) [inline src/langchain_google_memorystore_redis/doc_saver.py:72]: page_content shouldn't be encoded in json. we are supposed to save it in plain text

File: src/langchain_google_memorystore_redis/doc_saver.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                key is serialized by JSON.
+        """
+        if not metadata:
+            return {}
+        filtered_fields = (
+            self._metadata_fields & metadata.keys()
+            if self._metadata_fields
+            else metadata.keys()
+        )
+        filtered_metadata = {k: json.dumps(metadata[k]) for k in filtered_fields}
```
Comment (PingXie) [inline src/langchain_google_memorystore_redis/doc_saver.py:96]: I think the consensus is to save the top level field as-is and only encode the field if it is a "composite" structure. please check out "add_texts" in the vector store integration for references/alignment

File: src/langchain_google_memorystore_redis/vector_store.py [hunk 5]
```diff
@@ -408,7 +411,6 @@ def add_texts(
             zip_longest(keys_or_ids, texts, embeddings, metadatas), start=1
         ):
             key, text, embedding, metadata = bundle
-            key = self.key_prefix + key
```
Comment (PingXie) [inline src/langchain_google_memorystore_redis/vector_store.py:411]: why is this line removed?

File: src/langchain_google_memorystore_redis/doc_saver.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            doc_ids = [self._key_prefix + str(uuid.uuid4()) for _ in documents]
+        if len(documents) != len(doc_ids):
+            raise ValueError("The length of documents must match the length of the IDs")
+
+        for i, doc in enumerate(documents):
+            mapping = self._filter_metadata_by_fields(doc.metadata)
+            mapping.update({self._content_field: doc.page_content})
+
+            # Remove existing key in Redis to avoid reusing the doc ID.
+            self._redis.delete(doc_ids[i])
```
Comment (PingXie) [inline src/langchain_google_memorystore_redis/doc_saver.py:76]: should we consider pipelining as well?

File: tests/test_doc_saver.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        ),
+        (
+            '"content3"',
+            {"key1": {"k": "not_in_filter"}, "key2": {'"key"': "in_filter"}},
+            "page_content",
+            set(["key2"]),
+        ),
+    ],
+)
+def test_doc_saver_add_documents_one_doc(
```
Comment (PingXie) [inline tests/test_doc_saver.py:49]: curious - who calls this function?

File: src/langchain_google_memorystore_redis/vector_store.py (hunk 5 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/vector_store.py:411]: Because prefix is inserted on L401

File: tests/test_doc_saver.py (hunk 7 above)
Comment (craig08) [reply tests/test_doc_saver.py:49]: The integration test [does](https://github.com/googleapis/langchain-google-memorystore-redis-python/blob/staging/integration.cloudbuild.yaml#L21-L26). You can also invoke it by `REDIS_URL=redis://localhost:6379 python -m pytest` if you have a redis server listening on localhost:6379

File: src/langchain_google_memorystore_redis/doc_saver.py (hunk 6 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/doc_saver.py:76]: Can you elaborate more about pipelining?

File: src/langchain_google_memorystore_redis/vector_store.py (hunk 5 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/vector_store.py]: Sorry, read again and found out that if user specify keys_or_ids, the prefix is indeed removed. Added it back in the next commit.

File: src/langchain_google_memorystore_redis/doc_saver.py (hunk 6 above)
Comment (PingXie) [reply src/langchain_google_memorystore_redis/doc_saver.py:76]: see https://github.com/googleapis/langchain-google-memorystore-redis-python/blob/1097c08aff5dda4989672271af64f4dd8c3c634d/src/langchain_google_memorystore_redis/vector_store.py#L406

Comment (craig08) [reply src/langchain_google_memorystore_redis/doc_saver.py]: Thanks, updated.

**Review by PingXie — APPROVED** [review]
LGTM
